docs: improve analyze guide (#2910)
biomejs/biome: crates/biome_analyze/CONTRIBUTING.md

### Create and implement the rule

> New rules **must** be placed inside the `nursery` group. This group is meant as an incubation space, exempt from semantic versioning. Once a rule is stable, it's promoted to a group that fits it. This is done a minor/major release.
> 
> > [!TIP]
> > As developer, you aren't forced to make a rule perfect in one PR. Instead, you are encouraged to lay off a plan and to split the work in multiple PRs.
> >
> > If aren't familiar with Biome's APIs, this is an option that you have. If you decided to use this option, you should make sure to describe your plan in an issue.

Reviewer: nits;
```suggestion
New rules **must** be placed inside the `nursery` group. This group is meant as an incubation space, exempt from semantic versioning. Once a rule is stable, it's promoted to a group that fits it. This is done in a minor/major release.

> [!TIP]
> As a developer, you aren't forced to make a rule perfect in one PR. Instead, you are encouraged to lay off a plan and to split the work into multiple PRs.
>
> If you aren't familiar with Biome's APIs, this is an option that you have. If you decide to use this option, you should make sure to describe your plan in an issue.
```

Reviewer 2: Plus a tiny bit: `lay off a plan` => `lay out a plan`

}
```

### Coding the rule

Below, there are many tips and guidelines on how to create a lint rule using Biome infrastructure.

> #### Navigating the CST

Reviewer 3: We previously used the terminology `AST`. We should stick with this terminology?
```suggestion
#### Navigating the AST
```

Author: I think we should change to CST (maybe in another PR)

> Then navigating the nodes and tokens of certain nodes, will you notice straight away that the majority of those methods will return a `Result` (`SyntaxResult`).

Reviewer 3: ```suggestion
Then navigating the nodes and tokens of certain nodes, you will notice straight away that the majority of those methods will return a `Result` (`SyntaxResult`).
```

> Generally, you will end up navigating the CST inside the `run` function, and this function will usually return an `Option` or a `Vec`.

Reviewer 3: ```suggestion
Generally, you will end up navigating the AST inside the `run` function, and this function will usually return an `Option` or a `Vec`.
```

- If the `run` function returns an `Option`, you're encouraged to transform `Result` into `Option` and use the try operator `?`. This will make your coding way easier:
  ```rust
  fn run() -> Self::Signals {
    let prev_val = js_object_member.value().ok()?;
  }
  ```
- If the `run` function returns a `Vec`, you're encouraged to use the `let else` trick to reduce code branching:

  ```rust
    fn run() -> Self::Signals {
      let Ok(prev_val) = js_object_member.value() else {
        return vec![]
      };
    }
  ```

#### Query multiple nodes

> There are times when you might need to query multiple nodes at the time. Instead of querying the root of the CST, you can use the macro `declare_node_union!` to "join" multiple nodes into an `enum`:

Reviewer 3: ```suggestion
There are times when you might need to query multiple nodes at once. Instead of querying the root of the AST, you can use the macro `declare_node_union!` to "join" multiple nodes into an `enum`:
```

use biome_js_syntax::{AnyJsFunction, JsMethodObjectMember, JsMethodClassMember};

declare_node_union! {
  pub AnyFunctionLike = AnyJsFunction | JsMethodObjectMember | JsMethodClassMember
}
```

When creating a new node like this, we internally prefix them with `Any*` and postfix them with `*Like`. This is our internal naming convention.

The type `AnyFunctionLike` implements the trait `AstNode`, which means that it implements all methods such as `syntax`, `children`, etc.

#### `declare_rule`

This macro is used to declare an analyzer rule type, and implement the [RuleMeta] trait for it.

The macro itself expect the following syntax:

```rust,ignore
use biome_analyze::declare_rule;

declare_rule! {
    /// Disallow the use of `var`.
    ///
    /// ## Examples
    ///
    /// ### Invalid
    ///
    /// ```js,expect_diagnostic
    /// var a, b;
    /// ```
    pub(crate) NoVar {
        version: "1.0.0",
        name: "noVar",
    /// Documentation
    pub(crate) ExampleRule {
        version: "next",
        name: "myRuleName",
        language: "js",
        deprecated: "Use the rule `noAnotherVar`",
        recommended: false,
    }
}
```

> ### Custom Visitors

Reviewer 3: Should we add some words about `fix_kind` and `sources`?

    }
}

impl Rule for ExampleRule {
    fn diagnostic(ctx: &RuleContext<Self>, _state: &Self::State) -> Option<RuleDiagnostic> {
        Some(RuleDiagnostic::new(
            rule_category!(),
            ctx.query().text_trimmed_range(),
            "message",
        ))
    }
}
```

#### Semantic Model

> The semantic model provides information about the references of a binding (variable) within a program, indicating if it is written (e.g., `const a = 4`), read (e.g., `const b = a`, where `a` is read), or exported.

Reviewer 3: ```suggestion
The semantic model provides information about the references of a binding (declaration) within a program, indicating if it is written (e.g., `const a = 4`), read (e.g., `const b = a`, where `a` is read), or exported.
```
